Deployment.md

```diff
@@ -0,0 +1,33 @@
+# Deployment Practices
+
+Most of our Python projects (ListenBrainz, AcousticBrainz and CritiqueBrainz) have beta instances
+running alongside the production instances. The production instances follow the `production`
+branch in the git repository, while the beta instances follow the `master` branch.
+Most code (excluding very urgent hotfixes) is first merged into `master` and deployed into
+beta before being officially released into production. Urgent hotfixes are generally fixed by
+pull requests against the `production` branch directly.
+
+# Steps to deploy to production
+
+* Merge `master` into `production`, most of the code should already have been reviewed, but we still
+prefer opening Pull Requests for this merge before release.
+* Create a git tag of the format `v-$YEAR-$MONTH-$DATE.$VERSION`, for example `v-2019-07-04.0`.
+* Push to GitHub.
+* Deploy the `production` branch. This involves a few general steps that are similar for all projects. See example script below.
+
+```
+
+# pull latest master branch
+git checkout master && git pull origin master
+
+# push new image to docker hub
+# AcousticBrainz, CritiqueBrainz and ListenBrainz have a helper `push.sh` script
+# in the `docker` directory that we use.
+./push.sh prod v-2019-07-05.0
```

Comment:
Is this number the same as the tag that was used to tag git? I understand that this is only used to tag the docker image. Does it make sense to somehow use this to also check out the git tag?

What directory do you run this in? If it's your local development folder, then there's a risk that you have local changes that might be added to the image. `git describe --tags --dirty` will have the -dirty flag too. It'd be good to recommend that we have a separate build directory to do this from. Ideally we could build and push this image from CI when a git tag is written.

There's a dangerous step here where we pass a build argument for the deploy env based on the first arg to push.sh, but this value isn't reflected in the image name. This means we might push `acousticbrainz:v1-2-3.0` with beta startup scripts, and then accidentally run it as the prod container. Will this cause problems? Is this related to https://tickets.metabrainz.org/projects/AB/issues/AB-408? Do other python brainzes have this same problem?